src/guides/metamask-snap/index.md

```diff
@@ -0,0 +1,1318 @@
+
+```javascript
+const transactionData = remove0x(transaction.data);
+
+// Get function signature, i.e., the first 4 bytes of the data.
+const functionSignature = transactionData.slice(0, 8);
+
+let matchingFunction = '';  
+
+switch (functionSignature) {
+  case '4e1ca120':
+    matchingFunction = 'approveWithdraw(address,uint256)';
+    break;
+  case '97be5523':
+    matchingFunction = 'depositNFT(address,uint256,address)';
+    break;
+  case 'b537b269':
+    matchingFunction = 'removeApproval(address,uint256)';
+    break;
+  case '6088e93a':
+    matchingFunction = 'withdrawNFT(address,uint256)';
+    break;
+  default:
+    break;
+}
+
+if(matchingFunction.length > 0) { 
+  returnObject.method = matchingFunction; 
+}
+```
+
+After you reconnect your snap, you should see something like this:
+
+![function insight](./img/function_insight.gif)
+
+To get even more granular, let's actually decode the argument data when `withdrawNFT` is called. First, add this function, which will convert our data to strings:
+
+```javascript
+/**
+ * The ABI decoder returns certain which are not JSON serializable. This
+ * function converts them to strings.
+ *
+ * @param value - The value to convert.
+ * @returns The converted value.
+ */
+function normalize4ByteValue(value: unknown): unknown {
+  if (Array.isArray(value)) {
+    return value.map(normalize4ByteValue);
+  }
+
+  if (value instanceof Uint8Array) {
+    return bytesToHex(value);
+  }
+
+  if (typeof value === 'bigint') {
+    return value.toString();
+  }
+
+  return value;
+}
+```
+
+Then, in the `NFTVault` switch case, add:
+
+```javascript
+if(matchingFunction === 'withdrawNFT(address,uint256)') { 
+
+  // This is a function name in the shape "functionName(arg1Type,arg2Type,...)", so
+  // we do a simple slice to get the argument types.
+  const parameterTypes = matchingFunction
+    .slice(matchingFunction.indexOf('(') + 1, matchingFunction.indexOf(')'))
+    .split(',');
+
+  // Decode the parameters using the ABI utils library.
+  const decodedParameters = decode(
+    parameterTypes,
+    add0x(transactionData.slice(8)),
+  );
+
+  returnObject.args = decodedParameters.map(normalize4ByteValue); 
+
+  // now show them whether they are approved to withdraw or not
+  returnObject.canWithdraw = 'No'; 
+  try {
+    const provider = new ethers.providers.Web3Provider(wallet); 
+    const vaultContract = new ethers.Contract(
+      NFTVaultContractAddress,
+      NFTVault.abi,
+      provider,
+    );
+
+    const ethersReadResult = await vaultContract.getApproval(...returnObject.args);
+    if (ethersReadResult.length === 3 && ethersReadResult[2] === true) {
+      returnObject.canWithdraw = 'Yes';
+    }
+    returnObject.readResult = ethersReadResult; 
+
+  } catch (err) {
+    returnObject.canWithdraw = `${err}`;
+  }
+}
+```
+
+Now, when you call `withdrawNFT`, the snap should give you some information about whether or not the NFT can be drawn and who the owner and second signer are!
```

Comment:
can be **with**drawn

Reply:
Also worthwhile to note that this function is reading data from the chain, using the getApproval method we wrote in NFTVault.sol, and we can do that because withdrawNFT has the same arguments that we need for getApproval. So we are getting the arguments used for withdrawNFT and passing those to the onchain contract getApproval method to see if the NFT is approved for withdrawal or not. 

Plus, as a bonus, the Snap could be updated to use this information: 

- To tell the user the address of the second signer
- To tell them if they are attempting to withdraw an NFT they do not own
- To tell them if the NFT is not found (the first two values will be 0x0)